### PR TITLE
Add 32-bit PowerPC support

### PR DESCRIPTION
**Pull request by jamcleod**
Known possible issues:

* `gdb-multiarch` includes SIMD in its `powerpc:common` definition, which is what I based this off of. This seems to be different than the binutils source you link to as references, but I can't actually recreate the behavior on any GDB versions I have my hands on (including a native PowerPC one that I believe corresponds to Debian 7.3(?)).
* What I called `PowerPcCoreRegs` more so corresponds to `powerpc:common`. I considered splitting it up into core, fpu, and altivec because composability and whatnot, however that didn't really match with the structures for other architectures. A possible (low-effort) solution is just renaming it to `PowerPcCommonRegs`, as wordy as it is.
* I added some round-trip tests as I found those helpful while developing, but I don't actually know if you have any test style preferences this goes against

As per the past few architectures, just let me know if you have any issues. Apologies if the history is a bit of a mess, hopefully squash before merging should fix that.

### CODE REVIEW

File: src/arch/ppc/reg/mod.rs
Comment (daniel5151) [inline src/arch/ppc/reg/mod.rs:50]: (this applies to all similar implementations)

Ah, if only Rust had ergonomic array concatenation / safe transmutation...

Here's an alternative implementation:

```rust
u128::from_be_bytes()
```

Taking a look at godbolt, release-level ends up being significantly shorter than your implementation, though debug builds will take a performance hit due to the additional iterator + bounds checking...

https://godbolt.org/z/5xdr1b

As with my other comment below, it doesn't _really_ matter how this method is implemented, since it won't be on the hot-path anyways.

In this case, I actually think your approach, while "wordier", is the better way to go, even if it doesn't generate "optimal" assembly.

File: src/arch/ppc/reg/mod.rs
Comment (daniel5151) [inline src/arch/ppc/reg/mod.rs:156]: nit: just call it `mod tests`

**Review by daniel5151 — CHANGES_REQUESTED** [review]
Hey hey, another arch implementation! Awesome!

As always, I've got some nits and comments, but it mostly LGTM :+1: 

This particular set of changes includes some interesting digressions into one of Rust's pain points - namely, working with fixed-size arrays. I think you'll get a kick out of some of the interesting comparisons i've left in the comments :smile: 

I'm not too worried about the handwavey approach we've been taking wrt. architecture features, since it's something that requires a bit more involved plumbing work (that I haven't gotten around to). When I figure out a Good Way to control target features, I'll revisit the various arch implementations and manually clean things up.

And yeah, don't worry about the messy commit history. I'll just squash before merging :+1:

File: src/arch/ppc/mod.rs
Comment (jamcleod) [reply src/arch/ppc/mod.rs:7]: Heh, a relic from when I was mislead into thinking my target was 64-bit and being really confused why everything was 32-bit! Good eye.

**Comment by daniel5151** [comment]
Wait a minute, now that I think about it, I'm not sure there's any reason to define a dedicated `PpcVector` type in `gdbstub` itself. 

`gdbstub` aught to Keep Thing Simple, and just expose the register as a `u128` (or `[u8; 16]`), and not worry about the details of how those bytes should be reinterpreted.

I think it would be best to remove `PpcVector` from `gdbstub`, as it's not really related to the core functionality of the crate, and would add a bunch of code that would have to be maintained. Sorry!

File: src/arch/ppc/reg/core32.rs
Comment (jamcleod) [reply src/arch/ppc/reg/core32.rs:11]: Are you sure of that? That seems to include altivec+vsx, while mine is altivec (no vsx), vsx is a set of additional SIMD registers that aren't present in mine. Although your point stands, albeit with https://github.com/bminor/binutils-gdb/blob/master/gdb/features/rs6000/powerpc-altivec32.xml perhaps?

Comment (daniel5151) [reply src/arch/ppc/reg/core32.rs:11]: Oh, I didn't see that one, whoops :)

File: src/arch/ppc/mod.rs
Comment (jamcleod) [reply src/arch/ppc/mod.rs:9]: As per my other comment, maybe `PowerPcAltivec`?

Comment (daniel5151) [reply src/arch/ppc/mod.rs:9]: `PowerPcAltivec32`?

File: src/arch/ppc/reg/core32.rs
Comment (jamcleod) [reply src/arch/ppc/reg/core32.rs:85]: ah much better, not sure how that slipped my mind on this one!

**Comment by jamcleod** [comment]
> Wait a minute, now that I think about it, I'm not sure there's any reason to define a dedicated `PpcVector` type in `gdbstub` itself.
> 
> `gdbstub` aught to Keep Thing Simple, and just expose the register as a `u128` (or `[u8; 16]`), and not worry about the details of how those bytes should be reinterpreted.
> 
> I think it would be best to remove `PpcVector` from `gdbstub`, as it's not really related to the core functionality of the crate, and would add a bunch of code that would have to be maintained. Sorry!

I suppose so. My only real reasoning for this inclusion is that the conversion to and from is rather error prone, especially when you include the fact that GDB stores registers as little endian when communicating over a remote connection, PowerPC is (usually) big endian, and the actual interactions with this type will almost always require conversion between these types. This really gets tricky when it comes to the fact that the layout is endian dependent and cannot actually be represented as flat bytes.

I think a part of the problem with shoving this complexity on the user is that, when I initially implemented it as a type alias for `[u8; 16]`, it felt like it had a bit of a code-smell in that you had to have a bit of awareness about how GDB specifically handles the register which... felt a bit internal compared to the rest of the API gdbstub exposes? And I'd like to think only using the newtype for ensuring proper conversion/type safety doesn't seem like it'd cause too much of a maintenance burden (after all, the conversion remains the same, the end result is just a u128 that can be written raw for GDB).

I do see your point though, 150 lines of code just for a single vector type in a single architecture is pretty hefty. Ultimately, your call, if you'd rather I change it to just be a type alias for u128 then no sweat.

File: src/arch/ppc/mod.rs
Comment (jamcleod) [reply src/arch/ppc/mod.rs:9]: Sounds good!

**Comment by daniel5151** [comment]
> ...especially when you include the fact that GDB stores registers as little endian when communicating over a remote connection

Actually, that's not entirely right. The GDB documentation says the following (taken from the 'g' packet documentation): "The bytes with the register are transmitted in **target byte order**" It just so happens that all the architectures currently supported are little-endian, hence all the `to/from_le_bytes`.

Now, I'm certainly no power-pc expert, so this could be totally wrong, but most [bi-endian architectures](https://en.wikipedia.org/wiki/Endianness#Bi-endianness) are still fixed-endian within the processor (either little endian or big endian). i.e: even through data is fetched as big-endian from memory, the actual CPU register is loaded with it's little-endian equivalent. Again, I could be _totally wrong_ here, since I really haven't had too much experience with bi-endian systems.

As such, it's probably okay to use a u128 for the register, and as long as the internal de/serialization methods are consistent, and the user's de/serialization methods are consistent, everything should work out just fine, right?

In that case, you may want to update the `Registers` implementation to use `to/from_be_bytes`, if PowerPC is in-fact big-endian internally...

_Aside:_ Now that I think about it, there's a good chance that handling endianess properly is actually a pretty hairy issue, that might require some API rework... After all, `gdbstub` has only been tested on little-endian hosts + little-endian guests, so who knows what sorts of issues arise when it's run with a big-endian (or bi-endian) guest?

> 150 lines of code just for a single vector type in a single architecture is pretty hefty

...yeah, that's pretty much what it boils down to I'm afraid :disappointed:

Like I mentioned above, I have a feeling properly handling endianess will be something that'll require more exploration, and may even require tweaking part of `gdbstub`'s API a bit. For now, I'd like to err on the side of caution and keep the API as simple as possible (i.e: exposing a `[u8; 16]` or a `u128`), just so there's less code consider if/when the whole endianess issue is tackled...

**Comment by jam1garner** [comment]
> In that case, you may want to update the Registers implementation to use to/from_be_bytes, if PowerPC is in-fact big-endian internally...

Hm. That's rather unfortunate, especially given it doesn't seem possible to specify a system's byteorder from the target description. I honestly assumed that GDB defaulted to big endian for PowerPC (I mean, who wouldn't?) or at least check the MCR register to find the endianess... Oh well. Yeah I'll change the vectors to u128s and change it to big endian.

**Comment by jamcleod** [comment]
Changed it to big endian and changed the vector to a type alias for u128, I think that covers all the changes you've suggested so far.

File: src/arch/ppc/mod.rs
Comment (daniel5151) [inline src/arch/ppc/mod.rs:7]:
```suggestion
/// Implements `Arch` for 32-bit PowerPC + AltiVec SIMD Extensions
```

File: src/arch/ppc/reg/common.rs
Comment (daniel5151) [inline src/arch/ppc/reg/common.rs:6]: nit: capitalize FPU and properly case AltiVec

File: src/arch/ppc/reg/common.rs
Comment (daniel5151) [inline src/arch/ppc/reg/common.rs:70]: stale code (this is a noop)

**Review by daniel5151 — CHANGES_REQUESTED** [review]
Nice!

Just a couple of tiny nits (and a quick pass through `rustfmt`), and it should be set to merge :tada: 

Sorry again about the whole PpcVec thing! I'm glad we managed to work it out. 
And hey, on the bright side, it  it was a fun excuse to show off how magical the Rust compiler can be when optimizing release builds :smile:

**Comment by daniel5151** [comment]
Nice! And thanks again for all the PRs! 😄

**Pull request merged**
